Fix semantic of Queue.dequeueAll that was different with 2.12 strawman#530

wibisono opened the pull request:
The algorithm is a modification of Buffer.filterInPlace, storing elements matching predicate in new collection and keeping in place the one that does not match at the beginning of the queue, truncating the rest.

I also removed the work around on test/files/run/QueueTest.scala 

Fixes scala/collection-strawman#530

First non low hanging fruit that I attempted, if this is too far off the mark just let me know, I'll move on:P

wibisono commented:
Looking at this again, I wonder if I should've just used what was in the removed work around, to perform two passes, filtering for the result, and filtering in place for what's not dequeued.

szeiger reviewed (changes requested):
`IndexedOptimizedBuffer.filterInPlace` looks like a good starting place. I didn't even remember writing that code. Additionally one could optimize `filterInPlace` and `dequeueAll` further for `Queue` but the current implementation is reasonable and optimization can be left for another PR.

The scaladoc comment needs to be changed back to the 2.12 version though. It still reflects the changed semantics from the strawman.

wibisono commented:
Something I just noticed, I am not changing signature of the function from the previous strawman, but in 2.12, we were returning mutable.Seq instead of immutable.seq.

https://scastie.scala-lang.org/ACTRgbsMRE2ZcvXVSokTxw

Maybe I should also change the signature? then I will try to come up with some thing else.

szeiger commented:
Both are problematic. If code previously relied on getting a `mutable.Seq` it will break but if we kept returning a `mutable.Seq` any code that relied on getting a `collection.Seq` would break. My guess is that if you use mutable data structures you usually do so because of specific performance characteristics, so getting a `mutable.Seq` would not be very useful. I think we should return an `immutable.Seq` instead.

wibisono commented:
@szeiger  Thanks for the review/approve, I am a bit intrigued about possible optimization, perhaps I should've used partition from `StrictOptimizedIterableOps`
```
  def dequeueAll(p: A => Boolean): scala.collection.immutable.Seq[A] = {
    val (res, rem) = partition(p)
    this.clear
    this ++= rem
    res.toSeq
  }
```
I believe the partition will be faster since it is intended for iterable, but not sure about the next 3 lines.
This code also passed the test.

wibisono commented:
I am quite surprised with the benchmark result I did:
```
[info]
[info] Benchmark                         (size)  Mode  Cnt       Score      Error  Units
[info] QueueBenchmark.dequeueAPInteger       10  avgt   20     505.902 ±    8.651  ns/op
[info] QueueBenchmark.dequeueAPInteger     1000  avgt   20   39767.954 ±  819.477  ns/op
[info] QueueBenchmark.dequeueAPInteger    10000  avgt   20  436268.617 ± 2228.681  ns/op
[info] QueueBenchmark.dequeueAPString        10  avgt   20     487.725 ±    3.197  ns/op
[info] QueueBenchmark.dequeueAPString      1000  avgt   20   37389.763 ±  260.330  ns/op
[info] QueueBenchmark.dequeueAPString     10000  avgt   20  386099.096 ± 3358.049  ns/op
[info] QueueBenchmark.dequeueAllInteger      10  avgt   20     188.328 ±    1.682  ns/op
[info] QueueBenchmark.dequeueAllInteger    1000  avgt   20   16576.280 ±  100.398  ns/op
[info] QueueBenchmark.dequeueAllInteger   10000  avgt   20  166015.852 ± 8580.239  ns/op
[info] QueueBenchmark.dequeueAllString       10  avgt   20     173.278 ±    6.133  ns/op
[info] QueueBenchmark.dequeueAllString     1000  avgt   20   16542.336 ±  599.634  ns/op
[info] QueueBenchmark.dequeueAllString    10000  avgt   20  174675.009 ± 8657.415  ns/op
[success] Total time: 502 s, completed Apr 19, 2018 12:20:12 AM
```

The dequeueAP is the short version using partition, which seems much faster (2x more) 
Here is the benchmark commit  I did on a separate branch:
https://github.com/wibisono/scala/commit/c62c3e9ca7b40d9f76bac504f84d74d7cb184ad0

Before I switch to partition version, I need to make sure if I do anything wrong in benchmarking? or is it explainable that the filterInPlace based version was not actually faster than the simple partition based version?

Ichoran commented:
@wibisono - The numbers are nanoseconds, not operations/second, so lower numbers are better.  The "All" versions look faster to me.

wibisono commented:
You guys are right, so ill keep the current pr as it is. It was late night
benchmarking and I was used to see other benchmark from scalaz vs cats
effect that are all opsbper sec 😅

On Thu, 19 Apr 2018 at 15:01, Ichoran <notifications@github.com> wrote:

> @wibisono <https://github.com/wibisono> - The numbers are nanoseconds,
> not operations/second, so lower numbers are better. The "All" versions look
> faster to me.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/scala/scala/pull/6517#issuecomment-382725448>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAijg_MlbpoA9mZ7zyy2tDNKtTMHH1Xuks5tqIqfgaJpZM4TQkqS>
> .
>

adriaanm commented:
Please rebase when you get a chance.

wibisono committed:
Fix semantic of dequeueAll to be the same as version 2.12

The implementation is based on IndexedOptimizedBuffer.filterInPlace,
storing elements matching predicate in a new collection while keeping
elements not matching predicate at the beginning of the queue,
truncating the rest.

Remove the work around on test/files/run/QueueTest.scala

m

wibisono commented:
I rebased, and also in the previous PR I guess I made a mistake moving partest to JUnit, now I keep QueueTest where it is now in run/files just I remove the workaround before the fix.
